### PR TITLE
Add Unresponisve Node Up AND Responsive Partitioned Test

### PR DESCRIPTION
**Pull request by StrikingThirteen**
## Overview

Description:
Test an edge case in which not considering healing unresponsive nodes can turn a healing cluster to cluster which is down without the possibility of healing.

Why should this be merged: 
Tests that above scenario is dealt with correctly.

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                        ClusterStatusReport clusterStatusReport = corfuClient.getManagementView().getClusterStatus();
+                        // TODO: uncomment the following line after ClusterStatus API is fixed for partial partition
+                        // assertThat(clusterStatusReport.getClusterStatus()).isEqualTo(ClusterStatus.DEGRADED);
+                        // TODO: add node status check after we redefine NodeStatus semantics
+
+                        // Verify data path working fine
+                        for (int i = 0; i < DEFAULT_TABLE_ITER; i++) {
+                            assertThat(table.get(String.valueOf(i))).isEqualTo(String.valueOf(i));
+                        }
+                    });
```
[the remark this answers is not in the document]
Comment (StrikingThirteen) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: Done

File: it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                                                            .contains(server0.getEndpoint()),
+                                            corfuClient);
+
+                        // Step 4)
+                        assertThat(corfuClient.getLayout().getUnresponsiveServers())
+                                .contains(server0.getEndpoint());
+
+
+                        // Step 5)
+                        waitForUnresponsiveServersChange(size -> size == 1, corfuClient);
```
[the remark this answers is not in the document]
Comment (StrikingThirteen) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: There will be a time that there are potentially two unresponsive nodes before one of them getting healed. This step is to make sure that the one which could be healed is actually healed.

File: it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            }
+
+            testCase.it("Should fail the node with most link failures to unresponsive set",
+                    data -> {
+                        // Step 1)
+                        CorfuServer server0 = corfuCluster.getServerByIndex(0);
+                        CorfuServer server1 = corfuCluster.getServerByIndex(1);
+                        CorfuServer server2 = corfuCluster.getServerByIndex(2);
+
+                        // Step 2)
```
[the remark this answers is not in the document]
Comment (StrikingThirteen) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: Was trying to avoid repeating. Changed them to more expressive comments.

File: it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+                        assertThat(corfuClient.getLayout().getUnresponsiveServers())
+                                .containsExactly(server1.getEndpoint());
+
+                        // Step 3) Partition the responsive server0 from both unresponsive server1
+                        // and responsive server2 and reconnect server 1. Wait for layout's unresponsive
+                        // servers to change After this, cluster becomes unavailable.
+                        // NOTE: cannot use waitForClusterDown() since the partition only happens on server side, client
+                        // can still connect to two nodes, write to table so system down handler will not be triggered.
+                        server0.disconnect(Collections.singletonList(server1));
+                        server0.disconnect(Collections.singletonList(server2));
```
[the remark this answers is not in the document]
Comment (StrikingThirteen) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: Done.

File: it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+            CorfuClient corfuClient = corfuCluster.getLocalCorfuClient();
+
+            CorfuTable<String, String> table = corfuClient.createDefaultCorfuTable(DEFAULT_STREAM_NAME);
+            for (int i = 0; i < DEFAULT_TABLE_ITER; i++) {
+                table.put(String.valueOf(i), String.valueOf(i));
+            }
+
+            testCase.it("Should fail the node with most link failures to unresponsive set",
+                    data -> {
```
[the remark this answers is not in the document]
Comment (StrikingThirteen) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: Done.

File: it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (WenbinZhu) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: Oh I didn't notice you are using `contain` instead of `containsExactly`. But why don't we combine all these checks:

```
waitForLayoutChange(layout -> layout.getUnresponsiveServers()
                                       .contains(server0.getEndpoint()),
                                       corfuClient);

 // Verify server0 is unresponsive
assertThat(corfuClient.getLayout().getUnresponsiveServers())
                        .contains(server0.getEndpoint());

// Verify unresponsive server1 gets healed
waitForUnresponsiveServersChange(size -> size == 1, corfuClient);
assertThat(corfuClient.getLayout().getUnresponsiveServers())
                        .containsExactly(server0.getEndpoint());
``` 

into two: 
```
waitForUnresponsiveServersChange(size -> size == 1, corfuClient);
assertThat(corfuClient.getLayout().getUnresponsiveServers()).containsExactly(server0.getEndpoint());
```

which means  we wait until it stabilizes and do a final check?

[the remark this answers is not in the document]
Comment (StrikingThirteen) [reply it/src/test/java/org/corfudb/universe/scenario/NodeUpAndPartitionedIT.java]: Let's keep it as is so that in addition to checking the stabile result, we also assert that first step is adding the correct node to unresponsive set.
